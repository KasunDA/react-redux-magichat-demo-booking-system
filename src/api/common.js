export const delayedPromise = (value, delayInMS = 200) =>
    new Promise((resolve, _) => {
        setTimeout(
            () => resolve(value),
            delayInMS
        )
    });

export const failedPromise = (statusCode, status, reason, delayInMS = 200) =>
    new Promise((_, reject) => {
        setTimeout(
            () => reject({statusCode, status, reason}),
            delayInMS
        )
    });
