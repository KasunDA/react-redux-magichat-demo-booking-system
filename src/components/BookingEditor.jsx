import React from 'react'
import PropTypes from 'prop-types'
import {Formik, Field, Form} from 'formik'
import {
  ic_date_range,
  ic_group_work,
  ic_access_time,
  ic_face,
  ic_description,
  ic_done
} from 'react-icons-kit/md'

import {FormFieldWithIcon} from '../ui/forms'
import {getBookingObject, getInitialValues} from '../utilities/bookings'
import FieldTimePicker from './TimePicker'
import DateRangePicker from './DateRangePicker'

import {mergeDates} from '../utilities/dates'

class BookingEditor extends React.Component {
  static propTypes = {
    rooms: PropTypes.array.isRequired,
    bookings: PropTypes.array.isRequired,
    currentBooking: PropTypes.object.isRequired,
    onSaveBooking: PropTypes.func.isRequired
  }

  onSubmit = async (values, actions) => {
    actions.setStatus('')
    await this.props.onSaveBooking(getBookingObject(values))
    actions.setStatus('submitted')
  }

  validate = values => {
    let errors = {}

    if (!values.contact) {
      errors.contact = 'Required'
    }
    if (!values.startDate || !values.endDate) {
      errors.dates = 'Required'
    }
    if (!values.room || !values.room) {
      errors.room = 'Required'
    }

    return errors
  }

  render() {
    const {rooms, bookings, currentBooking} = this.props
    const booking = bookings.find(b => b.id === currentBooking.id)
    const initialValues = getInitialValues(booking, rooms)
    const {errors: bookingErrors = {}} = currentBooking

    return (
      <Formik
        enableReinitialize={true}
        validateOnChange={false}
        validateOnBlur={false}
        initialValues={initialValues}
        validate={this.validate}
        onSubmit={this.onSubmit}
        render={formProps => {
          const {values, touched, errors, status} = formProps

          const formIsSubmitted =
            !bookingErrors.isOverlapping && status === 'submitted'
              ? 'valid'
              : ''

          const normalizeEndTime = (value, values) => {
            const start = mergeDates(values.startDate, value)
            const end = mergeDates(values.endDate, values.endTime)
            if (start > end) {
              formProps.setFieldValue('endTime', value)
            }
          }

          return (
            <Form autoComplete="off">
              <FormFieldWithIcon
                icon={ic_group_work}
                error={touched.room && errors.room ? 'error' : ''}
                valid={formIsSubmitted}
              >
                <Field
                  component="select"
                  name="room"
                  className="form-control form-control-lg"
                >
                  <option label="Room" />
                  {rooms.map(room => (
                    <option key={room.id} value={room.id}>
                      {room.description}
                    </option>
                  ))}
                </Field>
              </FormFieldWithIcon>

              <FormFieldWithIcon
                icon={ic_date_range}
                error={
                  bookingErrors.isOverlapping ||
                  (touched.startDate && errors.dates)
                    ? 'error'
                    : ''
                }
                valid={formIsSubmitted}
              >
                <Field
                  name="dataRange"
                  render={() => (
                    <DateRangePicker
                      values={values}
                      onDatesChange={({startDate, endDate}) => {
                        formProps.setFieldValue('endDate', endDate)
                        formProps.setFieldValue('startDate', startDate)
                      }}
                    />
                  )}
                />
              </FormFieldWithIcon>

              {(values.startDate || values.endDate) && (
                <FormFieldWithIcon
                  icon={ic_access_time}
                  valid={formIsSubmitted}
                >
                  {values.startDate && (
                    <FieldTimePicker
                      name="startTime"
                      values={formProps.values}
                      onChange={value => {
                        formProps.setFieldValue('startTime', value)
                        normalizeEndTime(value, formProps.values)
                      }}
                    />
                  )}
                  {values.endDate && (
                    <FieldTimePicker
                      name="endTime"
                      values={formProps.values}
                      onChange={value => {
                        formProps.setFieldValue('endTime', value)
                      }}
                    />
                  )}
                </FormFieldWithIcon>
              )}

              <FormFieldWithIcon
                icon={ic_face}
                valid={formIsSubmitted}
                error={touched.contact && errors.contact ? 'error' : ''}
              >
                <Field
                  name="contact"
                  placeholder="Contact"
                  className="form-control form-control-lg"
                  data-testid="contact-input"
                />
              </FormFieldWithIcon>

              <FormFieldWithIcon icon={ic_description} valid={formIsSubmitted}>
                <Field
                  name="description"
                  render={({field}) => (
                    <textarea
                      className="form-control form-control-lg"
                      {...field}
                      placeholder="Description"
                    />
                  )}
                />
              </FormFieldWithIcon>

              <FormFieldWithIcon
                icon={formIsSubmitted ? ic_done : null}
                valid={formIsSubmitted}
              >
                <button
                  type="submit"
                  className="btn btn-primary"
                  data-testid="booking-submit"
                >
                  Submit
                </button>
              </FormFieldWithIcon>
            </Form>
          )
        }}
      />
    )
  }
}

export default BookingEditor
