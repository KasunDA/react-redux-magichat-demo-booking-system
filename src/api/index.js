import _ from 'lodash'
import roomIds from './data/roomIds.json'
import rooms from './data/rooms.json'
import bookingIds from './data/bookingIds.json'
import bookings from './data/bookings.json'
import {delayedPromise, failedPromise} from './common'
import uuid from 'uuid'
import Moment from 'moment'
import {extendMoment} from 'moment-range'

const moment = extendMoment(Moment)

export const getRoomIds = () => {
  // console.log(`getRoomIds()`)
  return delayedPromise(roomIds)
}

export const getRoom = roomId => {
  // console.log(`getRoom(${roomId})`)
  const room = _.get(rooms, roomId)
  if (_.isNil(room)) {
    return failedPromise(404, 'Not Found', `Room ${roomId} is Not Found`)
  } else {
    return delayedPromise(room)
  }
}

export const getRooms = async () => {
  try {
    const roomIds = await getRoomIds()
    const rooms = roomIds.map(async room => await getRoom(room))
    return Promise.all(rooms)
  } catch (error) {
    return failedPromise(400, 'Error', error)
  }
}

export const getBookingIds = () => {
  // console.log(`getBookingIds()`)
  // console.log(`bookingIds = ${JSON.stringify(bookingIds)}`)
  return delayedPromise(bookingIds)
}

export const getBooking = bookingId => {
  // console.log(`getBooking(${bookingId})`)
  const booking = _.get(bookings, bookingId)
  if (_.isNil(booking)) {
    return failedPromise(404, 'Not Found', `Booking ${bookingId} is Not Found`)
  } else {
    return delayedPromise(booking)
  }
}

export const getBookings = async () => {
  try {
    const bookingIds = await getBookingIds()
    const _bookings = bookingIds.map(async booking => await getBooking(booking))
    return Promise.all(_bookings)
  } catch (error) {
    return failedPromise(400, 'Error', error)
  }
}

/*
The following would be a possible API to perform a search on a server
*/
// export const filterBookings = async query => {
//   try {
//     const allBookings = await getBookings()

//     const filterBookingTexts = booking => {
//       const {roomId, contact, description} = booking
//       const bookingText = `${roomId} ${contact} ${description}`
//       return ~bookingText.toLowerCase().indexOf(query)
//     }

//     return query ? allBookings.filter(filterBookingTexts) : allBookings
//   } catch (error) {
//     return failedPromise(400, 'Error', error)
//   }
// }

export const cancelBooking = bookingId => {
  // console.log(`cancelBooking(${bookingId})`)
  return delayedPromise(() => {
    const booking = _.get(bookings, bookingId)
    if (_.isNil(booking)) {
      return failedPromise(
        404,
        'Not Found',
        `Booking ${bookingId} is Not Found`
      )
    } else {
      _.unset(bookings, bookingId)
      _.remove(bookingIds, bookingId)
      return delayedPromise(bookingId)
    }
  })
}

export const isOverlapping = async booking => {
  const allBookings = await getBookings()
  const filterCurrentIfExists = b => b.id !== booking.id
  const filterOtherRooms = b => booking.roomId === b.roomId
  const filterOverlappingBookings = b =>
    moment
      .range(b.start, b.end)
      .overlaps(moment.range(booking.start, booking.end))

  return (
    allBookings
      .filter(filterCurrentIfExists)
      .filter(filterOtherRooms)
      .filter(filterOverlappingBookings).length > 0
  )
}

export const setBooking = async booking => {
  const {id = uuid(), description, contact, roomId, start, end} = booking || {}
  const hasOverlappingBookings = await isOverlapping(booking)

  if (hasOverlappingBookings) {
    return failedPromise(400, 'Bad data', `Overlapping bookings`)
  } else {
    const newBooking = {id, description, contact, roomId, start, end}

    _.set(bookings, id, newBooking)
    _.union(bookingIds, id)

    return delayedPromise(newBooking)
  }
}
