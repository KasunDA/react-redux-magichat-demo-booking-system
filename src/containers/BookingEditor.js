import {connect} from 'react-redux'
import {onSaveBooking} from '../state/bookings/actions'
import BookingEditor from '../components/BookingEditor'

const mapStateToProps = state => ({
  rooms: state.rooms,
  bookings: state.bookings,
  currentBooking: state.currentBooking
})

const mapDispatchToProps = {
  onSaveBooking
}

const BookingItemContainer = connect(mapStateToProps, mapDispatchToProps)(
  BookingEditor
)

export default BookingItemContainer
