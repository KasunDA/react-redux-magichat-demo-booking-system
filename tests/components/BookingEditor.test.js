import 'react-dates/initialize'
import React from 'react'
import {render, Simulate} from 'react-testing-library'
import 'dom-testing-library/extend-expect'
import * as API from '../../src/api/index'
import BookingEditor from '../../src/components/BookingEditor'

describe('BookingEditor (Sample tests)', () => {
  it('should match the snapshot', async () => {
    const rooms = await API.getRooms()
    const bookings = await API.getBookings()

    const props = {
      rooms,
      bookings,
      currentBooking: {},
      onSaveBooking: jest.fn
    }

    const {container} = render(<BookingEditor {...props} />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should load the booking data in the Booking form', async () => {
    const rooms = await API.getRooms()
    const bookings = await API.getBookings()

    const props = {
      rooms,
      bookings,
      currentBooking: {id: bookings[0].id},
      onSaveBooking: jest.fn
    }

    const {getByTestId} = render(<BookingEditor {...props} />)
    const contactInput = getByTestId('contact-input')

    expect(contactInput.value).toEqual(bookings[0].contact)
  })

  it('should submit the form and we naively test that now', async () => {
    const rooms = await API.getRooms()
    const bookings = await API.getBookings()
    const onSaveBooking = jest.fn()
    const props = {
      rooms,
      bookings,
      currentBooking: {id: bookings[0].id},
      onSaveBooking
    }

    const {container} = render(<BookingEditor {...props} />)

    Simulate.submit(container.querySelector('form'))

    expect(onSaveBooking).toHaveBeenCalled()
  })
})
