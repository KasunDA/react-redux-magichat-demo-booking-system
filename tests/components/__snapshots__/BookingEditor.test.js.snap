// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingEditor (Sample tests) should match the snapshot 1`] = `
<form
  autocomplete="off"
>
  <div
    class="form-group sc-bwzfXH hBMOiY"
  >
    <div
      class="sc-bdVaJa EHgII"
      error=""
      fill="currentColor"
      size="24"
      style="display: inline-flex; justify-content: center; align-items: center;"
      valid=""
    >
      <svg
        fill="currentColor"
        height="24"
        style="display: inline-block; vertical-align: middle;"
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zM8 17.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5zM9.5 8c0-1.38 1.12-2.5 2.5-2.5s2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5S9.5 9.38 9.5 8zm6.5 9.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5z"
        />
      </svg>
    </div>
    <select
      class="form-control form-control-lg"
      name="room"
    >
      <option
        label="Room"
      />
      <option
        value="CAR"
      >
        Cardiff
      </option>
      <option
        value="DUB"
      >
        Dublin
      </option>
      <option
        value="EDN"
      >
        Edinburgh
      </option>
      <option
        value="LDN"
      >
        London
      </option>
    </select>
  </div>
  <div
    class="form-group sc-bwzfXH hBMOiY"
  >
    <div
      class="sc-bdVaJa EHgII"
      error=""
      fill="currentColor"
      size="24"
      style="display: inline-flex; justify-content: center; align-items: center;"
      valid=""
    >
      <svg
        fill="currentColor"
        height="24"
        style="display: inline-block; vertical-align: middle;"
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          d="M9 11H7v2h2v-2zm4 0h-2v2h2v-2zm4 0h-2v2h2v-2zm2-7h-1V2h-2v2H8V2H6v2H5c-1.11 0-1.99.9-1.99 2L3 20c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 16H5V9h14v11z"
        />
      </svg>
    </div>
    <div
      class="DateRangePicker DateRangePicker_1 DateRangePicker__block DateRangePicker__block_2"
    >
      <div>
        <div
          class="DateRangePickerInput DateRangePickerInput_1 DateRangePickerInput__withBorder DateRangePickerInput__withBorder_2 DateRangePickerInput__block DateRangePickerInput__block_3"
        >
          <div
            class="DateInput DateInput_1"
          >
            <input
              aria-describedby="DateInput__screen-reader-message-bookingStartDate"
              aria-label="Start Date"
              autocomplete="off"
              class="DateInput_input DateInput_input_1"
              id="bookingStartDate"
              name="bookingStartDate"
              placeholder="Start Date"
              type="text"
              value=""
            />
            <p
              class="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
              id="DateInput__screen-reader-message-bookingStartDate"
            >
              Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
            </p>
          </div>
          <div
            aria-hidden="true"
            class="DateRangePickerInput_arrow DateRangePickerInput_arrow_1"
            role="presentation"
          >
            <svg
              class="DateRangePickerInput_arrow_svg DateRangePickerInput_arrow_svg_1"
              viewBox="0 0 1000 1000"
            >
              <path
                d="M694.4 242.4l249.1 249.1c11 11 11 21 0 32L694.4 772.7c-5 5-10 7-16 7s-11-2-16-7c-11-11-11-21 0-32l210.1-210.1H67.1c-13 0-23-10-23-23s10-23 23-23h805.4L662.4 274.5c-21-21.1 11-53.1 32-32.1z"
              />
            </svg>
          </div>
          <div
            class="DateInput DateInput_1"
          >
            <input
              aria-describedby="DateInput__screen-reader-message-bookingEndDate"
              aria-label="End Date"
              autocomplete="off"
              class="DateInput_input DateInput_input_1"
              id="bookingEndDate"
              name="bookingEndDate"
              placeholder="End Date"
              type="text"
              value=""
            />
            <p
              class="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
              id="DateInput__screen-reader-message-bookingEndDate"
            >
              Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="form-group sc-bwzfXH hBMOiY"
  >
    <div
      class="sc-bdVaJa EHgII"
      error=""
      fill="currentColor"
      size="24"
      style="display: inline-flex; justify-content: center; align-items: center;"
      valid=""
    >
      <svg
        fill="currentColor"
        height="24"
        style="display: inline-block; vertical-align: middle;"
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          d="M9 11.75c-.69 0-1.25.56-1.25 1.25s.56 1.25 1.25 1.25 1.25-.56 1.25-1.25-.56-1.25-1.25-1.25zm6 0c-.69 0-1.25.56-1.25 1.25s.56 1.25 1.25 1.25 1.25-.56 1.25-1.25-.56-1.25-1.25-1.25zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8 0-.29.02-.58.05-.86 2.36-1.05 4.23-2.98 5.21-5.37C11.07 8.33 14.05 10 17.42 10c.78 0 1.53-.09 2.25-.26.21.71.33 1.47.33 2.26 0 4.41-3.59 8-8 8z"
        />
      </svg>
    </div>
    <input
      class="form-control form-control-lg"
      data-testid="contact-input"
      name="contact"
      placeholder="Contact"
      value=""
    />
  </div>
  <div
    class="form-group sc-bwzfXH hBMOiY"
  >
    <div
      class="sc-bdVaJa EHgII"
      fill="currentColor"
      size="24"
      style="display: inline-flex; justify-content: center; align-items: center;"
      valid=""
    >
      <svg
        fill="currentColor"
        height="24"
        style="display: inline-block; vertical-align: middle;"
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8l-6-6zm2 16H8v-2h8v2zm0-4H8v-2h8v2zm-3-5V3.5L18.5 9H13z"
        />
      </svg>
    </div>
    <textarea
      class="form-control form-control-lg"
      name="description"
      placeholder="Description"
    />
  </div>
  <div
    class="form-group sc-bwzfXH hBMOiY"
  >
    <span />
    <button
      class="btn btn-primary"
      data-testid="booking-submit"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
